src/tests/Microsoft.PowerFx.Core.Tests/TexlTests.cs

```diff
@@ -3981,6 +3981,65 @@ public void TestSingleColumnAccessDependentOnTableType(string script, string exp
             }
         }
 
+        [Theory]
+        [InlineData("ColumnNames(ParseJSON('Hello'))")]
+        [InlineData("ColumnNames(ParseJSON('[]'))")]
+        [InlineData("ColumnNames(ParseJSON('Hello').a)")]
+        [InlineData("ColumnNames(ParseJSON('{''a'':1}').a)")]
+        [InlineData("ColumnNames(Blank())")]
+        [InlineData("ColumnNames(ParseJSON('{''a'':{''b'':1}}').a)")]
+        public void TexlFunctionTypeSemanticsColumnNames(string expression)
+        {
+            var engine = new Engine(new PowerFxConfig());
+            var result = engine.Check(expression.Replace('\'', '\"'));
+
+            Assert.True(DType.TryParse("*[Value:s]", out var expectedDType));
+            Assert.Equal(expectedDType, result.Binding.ResultType);
+            Assert.True(result.IsSuccess);
+        }
+
+        [Theory]
+        [InlineData("ColumnNames({a:1,b:true})")] // Does not work with records
+        [InlineData("ColumnNames([1,2,3])")] // Does not work with arrays
+        [InlineData("ColumnNames(1)")]
+        public void TexlFunctionTypeSemanticsColumnNames_Negative(string expression)
+        {
+            var engine = new Engine(new PowerFxConfig());
+            var result = engine.Check(expression);
+
+            Assert.False(result.IsSuccess);
+        }
+
+        [Theory]
+        [InlineData("Column(ParseJSON('Hello'), 'a')")] // Compiles, would fail at runtime
+        [InlineData("Column(ParseJSON('[]'), 'Value')")] // Compiles, would fail at runtime
+        [InlineData("Column(ParseJSON('Hello').a, 'a')")] // Compiles, would fail at runtime
+        [InlineData("Column(ParseJSON('{''a'':1}'), 'a')")]
+        [InlineData("Column(ParseJSON('{''a'':1}').a, 'a')")] // Compiles, would fail at runtime
+        [InlineData("Column(Blank(), 'a')")]
+        [InlineData("Column(ParseJSON('{''a'':{''b'':1}}'), 'a')")]

```

Comment:
>{''b'':1 [](http://example.com/codeflow?start=46&length=8)

Test with nested column reference
ParseJSON('{''a'':{''b'':1}}').a , 'b'

Reply:
This is the same as the test above the Blank table (at least at compile time). I can add another runtime test for that.